### DISCUSSION TITLE
My feed occasionally misses posts, not sure why

### DISCUSSION THREAD

**Discussion by venteto** [MarshalX/atproto#243] [Questions] [answered]
I have a [pre-existing curated user list](https://bsky.app/profile/did:plc:6ho3kaxtbju3a5qbdfru7uae/lists/3kh356dnq5k2d) which was created directly in the Bluesky client, which just fetches posts from the Guggenheim, MoMA, and Tate art bots.

Now I have also [created a custom feed](https://bsky.app/profile/did:plc:6ho3kaxtbju3a5qbdfru7uae/feed/tb) to mimic the curated user list (and perhaps hopefully replace it). I can compare the output of my feed to that of the list, to see how well my feed is or is not working.

This is the Python snippet for my data filter for this specific feed:
```python
        museums = [
            'did:plc:jx26hj3zymmn53cfjetkmwry',  # guggenheim
            'did:plc:pa2epocjfar2fzz4symzmb5t'   # tate
            'did:plc:z5pxuiispc5yacvofktometn',  # moma
        ]
        if author_did in museums:
            post_dict = {
                'feed_name': os.environ.get('MU_SLUG', None),
                'uri': created_post['uri'],
                'cid': created_post['cid'],
            }
            posts_to_create.append(post_dict)
```

This is the extent of my gunicorn config file that is serving the app on Fly (I don't have the data stream indexer and the feeds server separated, I'm not sure how or if it would even make a difference):

```python
bind = '0.0.0.0:8000'
threads = 4
#workers = 1  # 2
#timeout = 120
```

But looking at the screenshot below, you can see my feed (on the right) is missing some of the posts caught by the curated list (on the left ... an example of a post my feed missed is circled in green):

![feed-missing-images](https://github.com/MarshalX/atproto/assets/7062385/fbc48448-5fa2-45bf-a6e9-70522fd16d9b)

It worries me that my feed is _already_ missing posts, before I even try to get it to compare post author DID's to a list of DID's stored in a _database_ table (for testing here, I just manually hard-coded three DID's in a Python list directly in the source code, which is not my long-term intent).

(Note: I'm not trying to resolve DID's in the Flask app, I have done that separately in an _offline_ Django app that is storing info for accounts ... ideally in the future I want the Flask app to be able to just _read_ from the separate Django SQLite database.)

**Comment by MarshalX** [comment]
Your provided code looks fine but `thread=4` could be problematic. There is no sense in listening to the network`s event 4 times. It's hard to say why it lost something... I can only guess:
1. you work with the firehose cursor poorly
2. your firehose client reconnects without restoring from the same place (about cursors too)
3. you have some race conditions because of thread count that could lead to some undefined behavior

**Reply by venteto** [reply]
I tried dropping it down to 1 thread and 1 worker but it started missing even more posts. Now I'm trying 2 workers and 2 threads, but it's already missed one post so far. I made [a repo here](https://github.com/venteto/bsky-feedgen) for the complete source code, if you wouldn't mind looking at it and pointing out any specific places I could improve, maybe how to run the firehose separately for example, if you think that would help me stop missing posts.

**Reply by MarshalX** [reply]
You can use this repo as example https://github.com/bluesky-astronomy/astronomy-feeds

**Comment by venteto** [comment, marked as answer]
I am closing this for now as it seems my Bluesky (bsky.app) [client language settings](https://github.com/bluesky-social/social-app/issues/2657#issuecomment-1913582625) are causing this problem
